### PR TITLE
GH-186 adding `OAuth2` Support to the `aem` bundle 

### PR DESCRIPTION
**Pull request by hayatkha**
The aem Workspace Supplier is missing the OAuth2 support (recently introduced in this sdk).
Added ability for com.adobe.aio.aem.workspace.internal.WorkspaceSupplierImpl to build an OAuthContext and pass it to the Workspace (adding the needed oauth config entries within com.adobe.aio.aem.workspace.ocd.WorkspaceConfig.
https://developer.adobe.com/developer-console/docs/guides/authentication/ServerToServerAuthentication/#oauth-server-to-server-credential

related issues: https://github.com/adobe/aio-lib-java/issues/186

### CODE REVIEW

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 1]
```diff
@@ -270,6 +280,10 @@ public Builder addMetascope(final String metascope) {
         jwtbuilder = JwtContext.builder();
       }
       jwtbuilder.addMetascope(metascope);
+      if(oAuthBuilder == null){
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:283]: this `addMetascope()` was only kept for jwt backward compatibility, let's not use it to set the oautscope ...

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 2]
```diff
@@ -314,10 +334,17 @@ public Workspace build() {
       if (authContext != null) {
         return new Workspace(imsUrl, imsOrgId, apiKey, consumerOrgId, projectId, workspaceId, authContext);
       }
+      if(oAuthBuilder == null) {
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:337]: I would recommend you do something like this instead:

```
 if (jwtbuilder != null) {
  return new Workspace(imsUrl, imsOrgId, apiKey, consumerOrgId, projectId, workspaceId, jwtbuilder.build());
 // to keep things backward compatible, jwt will be used first...
}
else  if (oauthBuilder != null) {
 return new Workspace(imsUrl, imsOrgId, apiKey, consumerOrgId, projectId, workspaceId, oAuthBuilder.build());
} 
else {
  throw new IllegalStateException("Missing auth confiugration, set either jwt or oauth...")
}
else
```

**Review by francoisledroff — CHANGES_REQUESTED** [review]
Thanks for your PR. here is a first quick review.
Could you rebase the PR against the new remote branch I just created: GH-186 (instead of `main` ) ?

File: aem/core_aem/src/main/java/com/adobe/aio/aem/workspace/ocd/WorkspaceConfig.java [hunk 3]
```diff
@@ -26,6 +26,10 @@
       description = "Comma separated list of metascopes associated with your API (`/s/event_receiver_api,/s/ent_adobeio_sdk` for instance) (project.workspace.details.credentials.jwt.meta_scopes)")
   String aio_meta_scopes() default "/s/ent_adobeio_sdk";
 
+  @AttributeDefinition(name = "OAuth Scopes",
+          description = "Comma separated list of oauth scopes associated with your API (project.workspace.details.credentials.oauth_server_to_server.scopes)")
```
Comment (francoisledroff) [inline aem/core_aem/src/main/java/com/adobe/aio/aem/workspace/ocd/WorkspaceConfig.java:30]: could you add a statement here in the `description` to inform this `
  String `  attribute is to be used for oauth only.
and similarly on line 25 could you upadte `Meta Scopes` to `JWT Meta Scopes`  
as well as the `description` to inform this `aio_meta_scopes` attribute is to be used for JWT only.

Comment (francoisledroff) [reply aem/core_aem/src/main/java/com/adobe/aio/aem/workspace/ocd/WorkspaceConfig.java:30]: could you also update https://github.com/adobe/aio-lib-java/tree/main/aem/core_aem#workspace-configuration accordingly ?

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 4]
```diff
@@ -254,6 +256,10 @@ public Builder clientSecret(final String clientSecret) {
         jwtbuilder = JwtContext.builder();
       }
       jwtbuilder.clientSecret(clientSecret);
+      if(oAuthBuilder == null){
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:259]: let's remove this new lines 259-262 ... 
clientSecret() method is to be deprecated and was only kept for jwt backward compatibilty 
we could mark is as such

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 1 above)
Comment (francoisledroff) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java:279]: please remove these new lines this only to be used for JWT

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 5]
```diff
@@ -314,10 +328,19 @@ public Workspace build() {
       if (authContext != null) {
         return new Workspace(imsUrl, imsOrgId, apiKey, consumerOrgId, projectId, workspaceId, authContext);
       }
+      if (oAuthBuilder == null) {
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:331]: let's remove line 331 to 336

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 6]
```diff
@@ -314,10 +328,19 @@ public Workspace build() {
...
+        oAuthBuilder = OAuthContext.builder();
+      }
       if (jwtbuilder == null) {
         jwtbuilder = JwtContext.builder();
       }
-      return new Workspace(imsUrl, imsOrgId, apiKey, consumerOrgId, projectId, workspaceId, jwtbuilder.build());
+      if (jwtbuilder != null && jwtbuilder.build().getPrivateKey() != null) {
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:337]: let's also remove `&& jwtbuilder.build().getPrivateKey() != null`

File: core/src/test/java/com/adobe/aio/workspace/WorkspaceTest.java [hunk 7]
```diff
@@ -18,6 +18,7 @@
 
 import com.adobe.aio.auth.Context;
 import com.adobe.aio.auth.JwtContext;
+import com.adobe.aio.auth.OAuthContext;
```
Comment (francoisledroff) [inline core/src/test/java/com/adobe/aio/workspace/WorkspaceTest.java:21]: let's remove this import

File: core/src/test/java/com/adobe/aio/auth/OAuthContextTest.java [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 public class OAuthContextTest {
 
-  private static final String TEST_PROPERTIES = "workspace.properties";
+  private static final String TEST_PROPERTIES = "workspace.oauth.properties";
```
Comment (francoisledroff) [inline core/src/test/java/com/adobe/aio/auth/OAuthContextTest.java:11]: you could also revert all your `resources` changes... right ?
they are not compulsory...

**Review by francoisledroff — CHANGES_REQUESTED** [review]
here is my second review:

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 1 above)
Comment (francoisledroff) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java:275]: let's remove line 275-277

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 9]
```diff
@@ -289,10 +294,14 @@ public Builder configMap(final Map<String, String> configMap) {
           .consumerOrgId(configMap.get(CONSUMER_ORG_ID))
           .projectId(configMap.get(PROJECT_ID))
           .workspaceId(configMap.get(WORKSPACE_ID));
-
       // For backwards compatibility - should this be kept?
-      jwtbuilder = JwtContext.builder();
-      jwtbuilder.configMap(configMap);
+      if(StringUtils.isBlank(configMap.get(OAuthContext.SCOPES))) {
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:298]:
```suggestion
      if(StringUtils.isNotBlank(configMap.get(JwtContext.META_SCOPES))) {
```

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 10]
```diff
@@ -289,10 +294,14 @@ public Builder configMap(final Map<String, String> configMap) {
...
+        jwtbuilder = JwtContext.builder();
+        jwtbuilder.configMap(configMap);
+      } else {
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:301]:
```suggestion
      } 
      if(StringUtils.isNotBlank(configMap.get(OAuthContext.SCOPES))) {
```

**Review by francoisledroff — CHANGES_REQUESTED** [review]
looking good, just a few more requests:

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 1 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

File: aem/core_aem/src/main/java/com/adobe/aio/aem/workspace/ocd/WorkspaceConfig.java (hunk 3 above)
Comment (hayatkha) [reply aem/core_aem/src/main/java/com/adobe/aio/aem/workspace/ocd/WorkspaceConfig.java]: Review Comments Incorporated

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 5 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

File: core/src/test/java/com/adobe/aio/workspace/WorkspaceTest.java (hunk 7 above)
Comment (hayatkha) [reply core/src/test/java/com/adobe/aio/workspace/WorkspaceTest.java]: Review Comments Incorporated

File: core/src/test/java/com/adobe/aio/auth/OAuthContextTest.java (hunk 8 above)
Comment (hayatkha) [reply core/src/test/java/com/adobe/aio/auth/OAuthContextTest.java]: Review Comments Incorporated

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 9 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 10 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 6 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 4 above)
Comment (hayatkha) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: Review Comments Incorporated

**Review by francoisledroff — COMMENTED** [review]
looking good to me..
we wil try to add intergration test around this for the final PR against `main`

**Review by shikhartanwar — APPROVED** [review]

**Review by francoisledroff — APPROVED** [review]

**Pull request merged**